### PR TITLE
use format option instead of value as a function

### PR DESCRIPTION
**Pull request by adamjimenez**

### CODE REVIEW

**Comment by deitch** [comment]
Is the purpose of this to separate `value` - which today handles both retrieving the data and presenting it - into 2 separate parts: `value`, which retrieves the raw value, and `format` which handles formatting?

What is the benefit of splitting it?

**Comment by adamjimenez** [comment]
Yes, that's the point. We need to reference the underlying value for sorting columns e.g. you can sort timestamps but not formatted dates. Plus it's useful for default column headers.

**Comment by deitch** [comment]
So the idea is that value will always be the raw value, and format gives formatting options? I have a vague memory of something else working like that. Something in Java Swing? Or Angular? It all blurs together after a while.

1. What happens if someone still has raw data in value, like `"$4.00"`; it still will work right? `value` is ignorant of what it has, it just takes the data as is?
2. If there is no format function, it just passes it straight to the cell unchanged?
3. What happens if `value` is a function? Do we keep backwards compatibility? Should we?
4. Look at my comment on the code for the commit.

File: README.md [hunk 1]
```diff
@@ -74,7 +74,8 @@ We use the div to control the entire height and width, and the span to get acces
 	* `columnClass`: a CSS class to add to the header cell and the column cell
 	* `cellClass`: a CSS class to add to each cell in this column (except for the header) - added to the <span>
 	* `wideCellClass`: a CSS class to add to each cell in this column (except for the header) - added to the <div>
-	* `value`: the attribute on the node to use as the value for this cell - entered as the <span> text. Can be a string or a function.
+	* `value`: the attribute on the node to use as the value for this cell - entered as the <span> text. Must be a string.
```
Comment (deitch) [inline README.md]: > Must be a string

Not really. It also can handle integers or other numbers, or even booleans, as you showed in the example. Probably should say "Must be a string, number, boolean or other primitive."

**Comment by adamjimenez** [comment]
1. If value is "price" and the attribute contains "$4.00", it will work.
2. Yes
3. It will display, but the column won't be sortable. I could put a `console.warn` in there.
4. Makes sense

**Comment by deitch** [comment]
So it will remain backwards compatible for `value` as a function? I have no issue with removing support for the function from the README.md, so no one new coming to it will see it and use it, but as long as we keep backwards support in there so no one needs to rewrite everything, that sounds fine.

Go for it, merge it in.

**Comment by adamjimenez** [comment]
updated readme and added a console warning.

**Comment by deitch** [comment]
Thanks. The only part that I still question is removing `value` as a function.

Based on our side discussion, I see why you need a unique column identifier, although as we saw, nothing prevents one from using the same `value` in 2 or 3 different columns.

But @jochenberger made a good case that for `value` to be a function if the input to sorting or formatting should be calculated. Think of the following example:

````javascript
var moneyFormat = function(inp) {
    return "$"+inp.toFixed(2);
},
extPrice = function(node){
    return node.data.price * node.data.size;
},
cols = columns: [
    {width: 110, header: "Nodes",title:"_DATA_"},
    {cellClass: "col1", value: "price", header: "Price", format: moneyFormat},
    {cellClass: "col2", value: "size", header: "Qty"},
    {cellClass: "col3", value: extPrice, header: "Extended Price", format: moneyFormat},
]
````

In the above case, the third column has *data* as the output of a function, not just the *format*.

**Comment by adamjimenez** [comment]
Another way is to change the data at source or add a loader function to jstree that adds the column data.

```
$('#tree').jstree({
    'core' : {
        'data' : {
            'url' : '/get/children/',
            'data' : function (node) {
                node.data.extPrice = node.data.price * node.data.size
                return node;
            }
        }
    });
```

**Comment by adamjimenez** [comment]
Also consider that by having value as a function then the sort mechanism would need to re-run that function for every single node when that column is sorted. Not terribly efficient.

**Comment by deitch** [comment]
> Also consider that by having value as a function then the sort mechanism would need to re-run that function multiple times for every node when that column is sorted. Not terribly efficient.

Are you saying that when we *render* it, it makes sense to rerun a function. But when we *sort* it, it doesn't make sense, since the value doesn't change?

Why would we need to rerun it for sorting? Are you saying it is necessary to retrieve the `value` with each sort, and therefore it either uses the reference in `value` to go to `node.data` to get it, or it has to run the `value` function?

Whereas the presentation is unchanged, so the `format` function need be run only on rendering?

> Another way is to change the data at source or add a loader function to jstree that adds the column data.

Let's think about this architecturally, see if it helps. What we are saying is the following. (Ignoring the "unique column reference" issue for now)

1. We need a way for a cell to have data that is both "plain" and "calculated". Call the output of this "raw" data. E.g. `25`, `"ABC"`, `10` ("calculated" = @jochenberger's point)
2. We need a way to "format", or change the presentation of the data. The input is "raw". Call the output of this "formatted" data. E.g. `25`, "User: ABC", `"$10.00"` ( @adamjimenez's  point about `format` fn)
3. Sorting should happen over "raw" data, whether plain or calculated. Sorting should never happen over formatted data. ( @adamjimenez's purpose in separating `format` from `value`)
4. "raw" data should be kept as is. It should never be changed or retrieved again, whether for sorting or any other purpose, unless the underlying data itself is changed. ( @adamjimenez's point in this comment)

@adamjimenez: your argument, I believe, is that the way it works, each sort goes back to get "raw" data via `value`. If `value` is a function, it will recalculate, *even though the data itself, and therefore the output of the fn, have not changed!*.

This means that we have 3 choices:

1. Leave it as is and accept the inefficiency of recalculating.
2. If `value` is a function, cache the output on the node, so sorting and other "display" activities go to the cache. Update the cache every time we run the function.
3. Remove `value` as a function, instead generating the "raw" data at a separate earlier stage.

I think that is what @adamjimenez is suggesting with the below. But I don't see how it works! Can you throw up a fiddle? Can I dynamically generate part of the data for a node in the tree?

> Another way is to change the data at source or add a loader function to jstree that adds the column data.

**Comment by adamjimenez** [comment]
@deitch The answer to all your questions is "yes". 
My suggestion is to prep the data before putting into the grid.
[Fiddle example](http://jsfiddle.net/adamjimenez/ycrz7grh/2/).
For ajax data you could use a jstree data function that does the same thing.

**Comment by deitch** [comment]
Take a look at the PR I flagged you on. It generates a unique ID for each column (except the tree, which remains `"text"`), and then references then. This way, fn `value` and primitive `value` both work.

Yes, it is inefficient to calculate the value multiple times if it is a function, but let's fix it and get it working first, then let's optimize.

**Comment by adamjimenez** [comment]
Generating random ids that can clash (no matter how unlikely) seems kludgey and indicative of the endemic nature of supporting this legacy feature. My aim would be to propose a fix that stops it breaking for now but to ultimately remove this option.

**Comment by deitch** [comment]
Doesn't seem so kludgy to me. You want to uniquely identify columns, use a unique identifier. I didn't want to use UUIDs - just way too big - but that is how so many things are done, from disk identifiers to user IDs, to whatever. But we can easily check. Commit in 2 mins.

**Comment by adamjimenez** [comment]
don't commit yet, still looking into this.

**Comment by deitch** [comment]
Is OK, is just going into the PR branch. take a look.

**Comment by jochenberger** [comment]
> indicative of the endemic nature of supporting this legacy feature
Just because you broke the functionality to add sorting, that doesn't necessarily make it a legacy feature.
And if one doesn't use the sort plugin, there is no need to write the data attributes anyway.
Maybe those two just aren't compatible. We could just disable sorting for columns with function-valued values for the moment.

**Comment by adamjimenez** [comment]
> Just because you broke the functionality to add sorting, that doesn't necessarily make it a legacy feature.

I'm not against it because I broke it. I'm against it because I think it's a bad idea and there are alternatives (which I've illustrated) that achieve the same thing without all the hacks.

> We could just disable sorting for columns with function-valued values for the moment.

I agree, and then ultimately deprecate this feature.

**Comment by adamjimenez** [comment]
see #161

**Comment by adamjimenez** [comment]
#162 resolves merge conflicts

**Pull request closed without merging**